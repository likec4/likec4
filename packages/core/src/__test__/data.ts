import { indexBy } from 'remeda'
import { ModelIndex } from '../model-index'
import type { Element, ElementKind, Fqn, Relation, RelationID, Tag } from '../types'

/**
              ┌──────────────────────────────────────────────────┐
              │                      cloud                       │
              │  ┌───────────────────────────────────────────┐   │
              │  │                 frontend                  │   │
┏━━━━━━━━━━┓  │  │   ┏━━━━━━━━━━━━━┓   ┏━━━━━━━━━━━━━━━━┓    │   │   ┏━━━━━━━━━━━┓
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┃ customer ┃──┼──┼──▶┃  dashboard  ┃   ┃   adminpanel   ┃◀───┼───┼───┃  support  ┃
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┗━━━━━━━━━━┛  │  │   ┗━━━━━━┳━━━━━━┛   ┗━━━━━━━━┳━━━━━━━┛    │   │   ┗━━━━━━━━━━━┛
              │  └──────────┼───────────────────┼────────────┘   │
              │             ├───────────────────┘                │
              │             │                                    │
              │  ┌──────────┼────────────────────────────────┐   │
              │  │          ▼       backend                  │   │
              │  │   ┏━━━━━━━━━━━━━┓       ┏━━━━━━━━━━━━━┓   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┃  graphlql   ┃──────▶┃   storage   ┃   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┗━━━━━━━━━━━━━┛       ┗━━━━━━┳━━━━━━┛   │   │
              │  └────────────────────────────────┼──────────┘   │
              └───────────────────────────────────┼──────────────┘
                                                  │
                                        ┌─────────┼─────────┐
                                        │ amazon  │         │
                                        │         ▼         │
                                        │ ┏━━━━━━━━━━━━━━┓  │
                                        │ ┃              ┃  │
                                        │ ┃      s3      ┃  │
                                        │ ┃              ┃  │
                                        │ ┗━━━━━━━━━━━━━━┛  │
                                        └───────────────────┘

specification {
  element actor
  element system
  element container
  element component

  tag old
}

model {

  actor customer
  actor support

  system cloud {
    container backend {
      component graphql
      component storage {
        #old
      }

      graphql -> storage
    }

    container frontend {
      component dashboard {
        -> graphql
      }
      component adminPanel {
        #old
        -> graphql
      }
    }
  }

  customer -> dashboard
  support -> adminPanel

  system amazon {
    component s3

    cloud.backend.storage -> s3
  }

}

 */
const el = ({
  id,
  kind,
  title,
  ...props
}: Partial<Omit<Element, 'id' | 'kind'>> & { id: string; kind: string }): Element => ({
  id: id as Fqn,
  kind: kind as ElementKind,
  title: title ?? id,
  description: null,
  technology: null,
  tags: null,
  links: null,
  ...props
})

export const fakeElements = {
  amazon: el({
    id: 'amazon',
    kind: 'system',
    title: 'amazon'
  }),
  cloud: el({
    id: 'cloud',
    kind: 'system',
    title: 'cloud'
  }),
  customer: el({
    id: 'customer',
    kind: 'actor',
    title: 'customer',
    shape: 'person'
  }),
  support: el({
    id: 'support',
    kind: 'actor',
    title: 'support',
    shape: 'person'
  }),
  'amazon.s3': el({
    id: 'amazon.s3',
    kind: 'component',
    title: 's3',
    shape: 'storage'
  }),
  'cloud.backend': el({
    id: 'cloud.backend',
    kind: 'container',
    title: 'backend'
  }),
  'cloud.frontend': el({
    id: 'cloud.frontend',
    kind: 'container',
    title: 'frontend',
    shape: 'browser'
  }),
  'cloud.backend.graphql': el({
    id: 'cloud.backend.graphql',
    kind: 'component',
    title: 'graphql'
  }),
  'cloud.backend.storage': el({
    id: 'cloud.backend.storage',
    kind: 'component',
    title: 'storage',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.adminPanel': el({
    id: 'cloud.frontend.adminPanel',
    kind: 'component',
    title: 'adminPanel',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.dashboard': el({
    id: 'cloud.frontend.dashboard',
    kind: 'component',
    title: 'dashboard'
  })
} satisfies Record<string, Element>

export type FakeElementIds = keyof typeof fakeElements

const rel = ({
  source,
  target,
  title
}: {
  source: FakeElementIds
  target: FakeElementIds
  title?: string
}): Relation => ({
  id: `${source}:${target}` as RelationID,
  title: title ?? '',
  source: source as Fqn,
  target: target as Fqn
})

export const fakeRelations = [
  rel({
    source: 'customer',
    target: 'cloud.frontend.dashboard'
  }),
  rel({
    source: 'support',
    target: 'cloud.frontend.adminPanel'
  }),
  rel({
    source: 'cloud.backend.storage',
    target: 'amazon.s3'
  }),
  rel({
    source: 'cloud.backend.graphql',
    target: 'cloud.backend.storage'
  }),
  rel({
    source: 'cloud.frontend.dashboard',
    target: 'cloud.backend.graphql'
  }),
  rel({
    source: 'cloud.frontend.adminPanel',
    target: 'cloud.backend.graphql'
  })
]

export type FakeRelationIds = keyof typeof fakeRelations

export const fakeModel = () =>
  ModelIndex.from({
    elements: fakeElements,
    relations: indexBy(fakeRelations, r => r.id)
  })
