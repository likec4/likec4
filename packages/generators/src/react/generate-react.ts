import type { DiagramView } from '@likec4/core'
import JSON5 from 'json5'
import { CompositeGeneratorNode, NL, expandToNode, joinToNode, toString } from 'langium'

const componentName = (value: string): string => {
  if (!value.charAt(0).match(/[a-zA-Z]/)) {
    value = 'View' + value
  }
  value = value.replaceAll('_', '')
  return value.charAt(0).toLocaleUpperCase() + value.slice(1)
}

const generateViewId = (views: DiagramView[]) =>
  joinToNode(views, view => expandToNode`'${view.id}'`, {
    separator: ' | '
  })

export function generateReact(views: DiagramView[]) {
  const components = views.map(({ id }) => {
    return {
      id,
      name: componentName(id)
    }
  })

  const out = new CompositeGeneratorNode()
  out.appendTemplate`
      /******************************************************************************
       * This file was generated
       * DO NOT EDIT MANUALLY!
       ******************************************************************************/
      /* eslint-disable */

      import type { DiagramView } from '@likec4/core'
      import { LikeC4 } from '@likec4/diagrams'
    `.append(NL, NL)

  if (components.length == 0) {
    out.append('export {}', NL)
    return toString(out)
  }

  out.appendTemplate`

      export type LikeC4ViewId = ${generateViewId(views)};
      export const LikeC4Views = {
    `
    .appendNewLine()
    .indent({
      indentation: 2,
      indentedChildren: indent => {
        indent.append(
          joinToNode(views, view => expandToNode`'${view.id}': (${JSON5.stringify(view)} as unknown) as DiagramView`, {
            separator: ',',
            appendNewLineIfNotEmpty: true
          })
        )
      }
    })
    .append('}  as const satisfies Record<LikeC4ViewId, DiagramView>', NL, NL).appendTemplate`
      export type LikeC4Views = typeof LikeC4Views

      export const {
        isViewId,
        useViewId,
        Diagram,
        Responsive,
        Fullscreen,
        Embedded,
        Browser,
      } = LikeC4.create<LikeC4ViewId>(LikeC4Views)

      export type DiagramProps = LikeC4.DiagramProps<LikeC4ViewId>
      export type ResponsiveProps = LikeC4.ResponsiveProps<LikeC4ViewId>
      export type FullscreenProps = LikeC4.FullscreenProps<LikeC4ViewId>
      export type EmbeddedProps = LikeC4.EmbeddedProps<LikeC4ViewId>
      export type BrowserProps = LikeC4.BrowserProps<LikeC4ViewId>

      // Re-export types
      export type {
        DiagramApi,
        Fqn,
        Element,
        RelationID,
        Relation,
        NodeId,
        EdgeId,
        ViewID,
        ComputedView,
        LikeC4Theme,
        DiagramNode,
        DiagramEdge,
        DiagramLabel,
        DiagramView
      } from '@likec4/diagrams'
    `.append(NL, NL)

  return toString(out)
}
