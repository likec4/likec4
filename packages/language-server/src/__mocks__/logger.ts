import { vi } from 'vitest'
import type { Logger } from '../logger'

export const logger = {
  debug: vi.fn(),
  info: vi.fn(),
  warn: vi.fn(),
  log: vi.fn(),
  error: vi.fn(),
  trace: vi.fn()
} satisfies Logger
