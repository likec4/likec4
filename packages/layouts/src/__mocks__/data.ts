import { computeView, assignNavigateTo } from '@likec4/core'
import { cloudView, cloud3levels, amazonView, fakeModel, indexView } from './model'

const model = fakeModel()

export const [
  computedIndexView,
  computedCloudView,
  computedCloud3levels,
  computedAmazonView
] = assignNavigateTo([
  computeView(indexView, model)!,
  computeView(cloudView, model)!,
  computeView(cloud3levels, model)!,
  computeView(amazonView, model)!
])
