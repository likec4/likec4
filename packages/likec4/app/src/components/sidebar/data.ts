import type { DiagramView } from '@likec4/core'
import type { TreeNodeData } from '@mantine/core'
import { useMemo } from 'react'
import { find, isTruthy, values } from 'remeda'
import { useLikeC4Views } from 'virtual:likec4/store'

interface DiagramTreeNodeData {
  label: string
  value: string
  type: 'file' | 'folder' | 'diagram'
  children: DiagramTreeNodeData[]
}

export const isTreeNodeData = (node: TreeNodeData): node is DiagramTreeNodeData =>
  'type' in node && ['file', 'folder', 'diagram'].includes(node.type as any)

function compareTreeNodes(a: DiagramTreeNodeData, b: DiagramTreeNodeData) {
  if (a.children.length === 0 && b.children.length > 0) {
    return 1
  }
  if (a.children.length > 0 && b.children.length === 0) {
    return -1
  }
  return a.label.localeCompare(b.label)
}

function buildDiagramTreeData(views: DiagramView[]): DiagramTreeNodeData[] {
  const root: DiagramTreeNodeData = {
    value: '',
    label: 'Diagrams',
    type: 'folder',
    children: []
  }

  const findParent = (path: string): DiagramTreeNodeData => {
    let parent = root
    if (path === '') {
      return parent
    }
    const segments = path.split('/')
    const traversed = ['@fs'] as string[]
    while (segments.length) {
      const label = segments.shift() as string
      traversed.push(label)
      const value = traversed.join('/')
      let node = find(parent.children!, n => n.value === value)
      if (!node) {
        node = { label, value, type: 'folder', children: [] }
        parent.children.push(node)
      }
      parent = node
    }
    return parent
  }

  for (const view of views) {
    const parent = findParent(view.relativePath ?? '')
    parent.children.push({
      value: view.id,
      label: view.title ?? view.id,
      type: 'diagram',
      children: []
    })
    if (parent !== root) {
      parent.children.sort(compareTreeNodes)
      if (isTruthy(view.relativePath) && parent.type !== 'file') {
        parent.type = 'file'
      }
    }
  }

  return root.children.sort(compareTreeNodes)
}

// const $diagramsTree = batched($views, views => buildDiagramTreeData(values(views)))

export function useDiagramsTreeData() {
  const views = useLikeC4Views() as unknown as Record<string, DiagramView>
  return useMemo(() => buildDiagramTreeData(values(views)), [views])
}
