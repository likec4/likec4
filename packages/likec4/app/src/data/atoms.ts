import type { ExtractAtomValue } from 'jotai'
import { atom } from 'jotai'
import { atomFamily, atomWithReducer, splitAtom } from 'jotai/utils'
import { equals, groupBy, mapObject, values } from 'rambdax'
import { LikeC4Views } from '~likec4'
import { buildDiagramTreeAtom } from './sidebar-diagram-tree'
import type { DiagramView } from '@likec4/core'

function atomWithCompare<Value>(initialValue: Value) {
  return atomWithReducer(initialValue, (prev: Value, next: Value) => {
    if (equals(prev, next)) {
      return prev
    }
    return next
  })
}

const likec4ViewFamily = atomFamily(
  (view: DiagramView) => {
    const atom = atomWithCompare(view)
    atom.debugLabel = `view#${view.id}`
    return atom
  },
  (a, b) => a.id === b.id
)

const mapToAtoms = (views: typeof LikeC4Views) => mapObject(view => likec4ViewFamily(view), views)

const _viewsAtom = atom(mapToAtoms(LikeC4Views))
_viewsAtom.debugLabel = '_views'

export const viewsAtom = atom(
  get => get(_viewsAtom),
  (_, set, update: typeof LikeC4Views) => {
    set(
      _viewsAtom,
      mapObject(view => {
        const atom = likec4ViewFamily(view)
        set(atom, view)
        return atom
      }, update)
    )
  }
)
viewsAtom.debugLabel = 'views'

const indexPageTilesAtom = atom(get => {
  const views = values(get(viewsAtom))
  const byPath = groupBy(v => get(v).relativePath ?? '', views)
  return Object.entries(byPath)
    .map(([path, views]) => ({
      path,
      isRoot: path === '',
      views
    }))
    .sort((a, b) => {
      return a.path.localeCompare(b.path)
    })
})
export type IndexPageTile = ExtractAtomValue<typeof indexPageTilesAtom>[number]

const byPath = (tile: IndexPageTile) => tile.path
export const indexPageTilesAtomsAtom = splitAtom(indexPageTilesAtom, byPath)

// export type DashboardTile = ExtractAtomValue<typeof dashboardTilesAtom>[number]

// const key = (tile: DashboardTile) => tile.path

// export const diagramsTreeAtom = atom(buildDiagramTreeAtom(LikeC4Views))
export const diagramsTreeAtom = atom(get => {
  const views = values(get(viewsAtom))
  return buildDiagramTreeAtom(views.map(v => get(v)))
})

export const selectLikeC4ViewAtom = (viewId: string) => {
  return atom(get => {
    const viewAtom = get(viewsAtom)[viewId]
    return viewAtom ? get(viewAtom) : null
  })
}

if (import.meta.hot) {
  let $updateViews: undefined | ((update: typeof LikeC4Views) => void)

  viewsAtom.onMount = set => {
    $updateViews = set
    return () => {
      $updateViews = undefined
    }
  }

  import.meta.hot.accept('/@vite-plugin-likec4/likec4-generated', md => {
    const update = md?.LikeC4Views as typeof LikeC4Views | undefined
    if (update) {
      $updateViews?.(update)
    }
  })
}
