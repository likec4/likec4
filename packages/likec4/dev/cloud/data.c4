model {

  extend cloud.data {

    hitsAnomalies = app 'Hits Anomalies' {
      description '
        Process page hits.
        Search for anomalies.
      '

      -> amazon.sqs.hits 'process hits data'
    }

  }

}

views {

  view cloud_data of cloud.data {
    title "[context] Data Services"
    include
        *,
        cloud,
        amazon,
        amazon.sqs,
        cloud.data -> amazon.sqs.*

    style cloud {
      color muted
    }
    style amazon.* {
      color secondary
    }
  }

}
