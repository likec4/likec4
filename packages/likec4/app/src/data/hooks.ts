import { useAtomValue } from 'jotai'
import { useMemo } from 'react'
import { diagramsTreeAtom, indexPageTilesAtomsAtom, selectLikeC4ViewAtom } from './atoms'

export const useLikeC4View = (viewId: string) => {
  const anAtom = useMemo(() => selectLikeC4ViewAtom(viewId), [viewId])
  return useAtomValue(anAtom)
}

export const useIndexPageTileAtoms = () => {
  return useAtomValue(indexPageTilesAtomsAtom)
}

export type { IndexPageTile } from './atoms'

export const useDiagramsTree = () => useAtomValue(diagramsTreeAtom)
