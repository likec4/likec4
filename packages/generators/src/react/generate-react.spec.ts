import { expect, test } from 'vitest'
import { fakeDiagram, fakeDiagram2 } from '../__mocks__/data'
import { generateReact } from './generate-react'

test('generate react', () => {
  expect(generateReact([fakeDiagram, fakeDiagram2])).toMatchSnapshot()
})
