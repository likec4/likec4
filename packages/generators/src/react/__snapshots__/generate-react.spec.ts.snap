// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate react 1`] = `
"/******************************************************************************
 * This file was generated
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/
/* eslint-disable */

import type { DiagramView } from '@likec4/core'
import { LikeC4 } from '@likec4/diagrams'


export type LikeC4ViewId = 'fakeView' | '74f5364fa4b';
export const LikeC4Views = {  
  'fakeView': ({edges:[{headArrow:[[294,487],[291,496],[288,487]],id:'cloud.frontend:cloud.backend',label:'requests',labelBox:{align:'left',width:62,x:295,y:459},points:[[291,406],[291,432],[291,461],[291,487]],relations:['cloud.frontend.dashboard:cloud.backend.graphql','cloud.frontend.adminPanel:cloud.backend.graphql'],source:'cloud.frontend',target:'cloud.backend'},{headArrow:[[294,701],[291,711],[288,701]],id:'cloud.backend:amazon',label:null,labelBox:null,points:[[291,641],[291,660],[291,681],[291,702]],relations:['cloud.backend.storage:amazon.s3'],source:'cloud.backend',target:'amazon'},{headArrow:[[344,257],[336,263],[338,253]],id:'support:cloud.frontend',label:null,labelBox:null,points:[[413,141],[391,176],[364,218],[342,255]],relations:['support:cloud.frontend.adminPanel'],source:'support',target:'cloud.frontend'},{headArrow:[[244,253],[246,263],[239,257]],id:'customer:cloud.frontend',label:'opens',labelBox:{align:'left',width:44,x:208,y:193},points:[[169,141],[191,176],[218,218],[241,255]],relations:['customer:cloud.frontend.dashboard'],source:'customer',target:'cloud.frontend'}],height:855,id:'fakeView',nodes:[{children:[],color:'primary',id:'amazon',parent:null,position:[165,715],relative:[165,715],shape:'rectangle',size:{height:140,width:251},title:'amazon'},{children:['cloud.backend','cloud.frontend'],color:'primary',id:'cloud',parent:null,position:[139,210],relative:[139,210],shape:'rectangle',size:{height:457,width:304},title:'cloud'},{children:[],color:'primary',id:'customer',parent:null,position:[0,0],relative:[0,0],shape:'rectangle',size:{height:140,width:251},title:'customer'},{children:[],color:'primary',id:'support',parent:null,position:[330,0],relative:[330,0],shape:'rectangle',size:{height:140,width:251},title:'support'},{children:[],color:'primary',id:'cloud.backend',parent:'cloud',position:[165,500],relative:[26,290],shape:'rectangle',size:{height:140,width:251},title:'backend'},{children:[],color:'primary',id:'cloud.frontend',parent:'cloud',position:[165,266],relative:[26,56],shape:'rectangle',size:{height:140,width:251},title:'frontend'}],rules:[{exprs:[{wildcard:true}],isInclude:true}],title:'',viewOf:'cloud',width:582} as unknown) as DiagramView,
  '74f5364fa4b': ({edges:[{headArrow:[[129,456],[126,465],[123,456]],id:'system.frontend:system.backend',label:'requests',labelBox:{align:'left',width:62,x:130,y:428},points:[[126,375],[126,400],[126,429],[126,456]],relations:['671e3db5eb605d5f245f247d95925bd0953e0e4c'],source:'system.frontend',target:'system.backend'},{headArrow:[[129,221],[126,231],[123,221]],id:'client:system.frontend',label:'opens',labelBox:{align:'left',width:44,x:130,y:193},points:[[126,141],[126,166],[126,195],[126,221]],relations:['99bffdfd5c78e024393e6223cd9b19225321a5e3'],source:'client',target:'system.frontend'}],height:609,id:'74f5364fa4b',nodes:[{children:[],color:'primary',id:'client',parent:null,position:[0,0],relative:[0,0],shape:'rectangle',size:{height:140,width:251},title:'client'},{children:[],color:'primary',id:'system.backend',parent:null,position:[0,469],relative:[0,469],shape:'rectangle',size:{height:140,width:251},title:'backend'},{children:[],color:'primary',id:'system.frontend',parent:null,position:[0,235],relative:[0,235],shape:'rectangle',size:{height:140,width:251},title:'frontend'}],rules:[{exprs:[{wildcard:true}],isInclude:true}],title:'frontend',viewOf:'system.frontend',width:251} as unknown) as DiagramView
} as const satisfies Record<LikeC4ViewId, DiagramView>

export type LikeC4Views = typeof LikeC4Views

export const {
  isViewId,
  useViewId,
  Diagram,
  Responsive,
  Fullscreen,
  Embedded,
  Browser,
} = LikeC4.create<LikeC4ViewId>(LikeC4Views)

export type DiagramProps = LikeC4.DiagramProps<LikeC4ViewId>
export type ResponsiveProps = LikeC4.ResponsiveProps<LikeC4ViewId>
export type FullscreenProps = LikeC4.FullscreenProps<LikeC4ViewId>
export type EmbeddedProps = LikeC4.EmbeddedProps<LikeC4ViewId>
export type BrowserProps = LikeC4.BrowserProps<LikeC4ViewId>

// Re-export types
export type {
  DiagramApi,
  Fqn,
  Element,
  RelationID,
  Relation,
  NodeId,
  EdgeId,
  ViewID,
  ComputedView,
  LikeC4Theme,
  DiagramNode,
  DiagramEdge,
  DiagramLabel,
  DiagramView
} from '@likec4/diagrams'

"
`;
